google/resource_storage_notification.go

```diff
@@ -0,0 +1,162 @@
+		Read:   resourceStorageNotificationRead,
+		Delete: resourceStorageNotificationDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceStorageNotificationImportState,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"payload_format": &schema.Schema{
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringInSlice([]string{"JSON_API_V1", "NONE"}, false),
+			},
+
+			"topic": &schema.Schema{
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
+			},
+
+			"custom_attributes": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"event_types": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{
+						"OBJECT_FINALIZE", "OBJECT_METADATA_UPDATE", "OBJECT_DELETE", "OBJECT_ARCHIVE"},
+						false),
+				},
+			},
+
+			"object_name_prefix": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"self_link": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceStorageNotificationCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	bucket := d.Get("bucket").(string)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	computedTopicName := getComputedTopicName(project, d.Get("topic").(string))
+
+	storageNotification := &storage.Notification{
+		CustomAttributes: expandStringMap(d, "custom_attributes"),
+		EventTypes:       convertStringSet(d.Get("event_types").(*schema.Set)),
+		ObjectNamePrefix: d.Get("object_name_prefix").(string),
+		PayloadFormat:    d.Get("payload_format").(string),
+		Topic:            computedTopicName,
+	}
+
+	res, err := config.clientStorage.Notifications.Insert(bucket, storageNotification).Do()
+	if err != nil {
+		return fmt.Errorf("Error creating notification config for bucket %s: %v", bucket, err)
+	}
+
+	d.SetId(fmt.Sprintf("%s/notificationConfigs/%s", bucket, res.Id))
+
+	return resourceStorageNotificationRead(d, meta)
+}
+
+func resourceStorageNotificationRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	bucket := d.Get("bucket").(string)
+	_, notificationID := resourceStorageNotificationParseID(d.Id())
+
+	res, err := config.clientStorage.Notifications.Get(bucket, notificationID).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Notification configuration %s for bucket %s", notificationID, bucket))
+	}
+
+	topic, _ := getRelativePath(res.Topic)
```

Comment:
Since we have the DiffSuppressFunc for self link or resource name, we should be fine just storing this the way it's returned from the API (either way is really fine, but since that's the case may as well go with the simpler solution)